### PR TITLE
Canavandl/4372 log color mapper

### LINKED ISSUE #4372: LogColorMapper?
Bokeh provides a ColorMapper base class and a [LinearColorMapper](http://bokeh.pydata.org/en/0.11.1/docs/reference/models/mappers.html), but no logarithmic color mapper as far as I can see.  Can a logarithmic mapper be provided?  matplotlib provides LogNorm and SymLogNorm, and it would be very useful if similar functionality were available in Bokeh.

### PR DESCRIPTION
**Pull request by canavandl**
- [x] issues: fixes #4372 
- [x] tests added / passed (needs more)
- [ ] release document entry (if new feature or API change)

Todo:
Currently doesn't handle instances where elements in image have value in the range (0, 1). Those pixels end up being black. Need to research what mpl does.

example code:
```python
import numpy as np
from matplotlib.mlab import bivariate_normal

from bokeh.plotting import figure, output_file, show
from bokeh.models import Range1d, LinearColorMapper, LogColorMapper

output_file('image.html')

N = 100
X, Y = np.mgrid[-3:3:complex(0, N), -2:2:complex(0, N)]

# A low hump with a spike coming out of the top right.  Needs to have
# z/colour axis on a log scale so we see both hump and spike.  linear
# scale only shows the spike.
Z1 = bivariate_normal(X, Y, 0.1, 0.2, 1.0, 1.0) +  \
    0.1 * bivariate_normal(X, Y, 1.0, 1.0, 0.0, 0.0)

image = Z1

mapper = LogColorMapper(palette="Spectral11")

p = figure(x_range=Range1d(0,1), y_range=Range1d(0,1))
p.image(image=[image], color_mapper=mapper,
        dh=[1.0], dw=[1.0], x=[0], y=[0])

show(p)
```

yields (sorry for the bad palette choice): 

screen shot 2016-06-14 at 1 15 02 pm

compared to LinearColorMapper:

screen shot 2016-06-13 at 7 25 38 pm

### CODE REVIEW

File: bokeh/models/glyphs.py [hunk 1]
```diff
@@ -353,7 +353,7 @@ def __init__(self, **kwargs):
     images to have a gap between them, when they should appear flush.
     """)
 
-    color_mapper = Instance(LinearColorMapper, help="""
+    color_mapper = Instance(ColorMapper, help="""
```
Comment (jbednar) [inline bokeh/models/glyphs.py]: Looks like this was a mistake before; good catch.

File: bokeh/models/mappers.py [hunk 2]
```diff
@@ -50,17 +50,47 @@ class LinearColorMapper(ColorMapper):
     this are clamped to ``high``.
     """)
 
-    # TODO: (jc) what is the color code for transparent?
-    # TODO: (bev) better docstring
-    reserve_color = Color("#ffffff", help="""
-    Used by Abstract Rendering.
```
Comment (jbednar) [inline bokeh/models/mappers.py]: I think we can safely remove anything to do with abstract rendering, but maybe that should be a separate PR?

Comment (bryevdv) [inline bokeh/models/mappers.py]: Agreed. And although it was added for AR it's potentially a useful feature in general. In any case removing properties should require a deprecation cycle at this point.

**Comment by jbednar** [comment]
> Currently doesn't handle instances where elements in image have value in the range (0, 1). Those pixels end up being black. Need to research what mpl does.

datashader uses log1p to avoid the issue, i.e. adding 1 to the value before taking the log.  mpl's code is a little tricky to figure out.  In matplotlib/colors.py, the LogNorm class does:

```python
            resdat = result.data
            mask = result.mask
            if mask is np.ma.nomask:
                mask = (resdat <= 0)
            else:
                mask |= resdat <= 0
            cbook._putmask(resdat, mask, 1)
            np.log(resdat, resdat)
            resdat -= np.log(vmin)
            resdat /= (np.log(vmax) - np.log(vmin))
            result = np.ma.array(resdat, mask=mask, copy=False)
```

**Comment by jbednar** [comment]
I'm not sure we need all the flexibility of changing the log base that's in there now.  Having lots of options probably actually hampers communication, since most people just want to say "logarithmic mapping -- check", not keep track of lots of details or have to communicate them to viewers of the data.  So I'd put in only one option for now unless there is pressure later to have multiple options and configurability for this feature.

**Comment by jbednar** [comment]
It's a bit discouraging that adding a new and relatively simple subclass requires changes to 10 different files. :-(  3 of those are for tests, which is thus just mirroring the fact that changes were spread over multiple files elsewhere.  And of course any project that is split across languages is going to need code at both ends, plus potentially glue between them.  Still, one would hope that one wouldn't have to update tiny little bits spread over *so* many different locations!

**Comment by bryevdv** [comment]
Well some of the work was to factor out code from a single class to a base class that could be shared. That's a one time cost

**Comment by jbednar** [comment]
Maybe it would have been good to separate that into two PRs, then, one for the refactoring (no change to behavior) and one for the new functionality.

File: bokeh/models/mappers.py (hunk 2 above)
Comment (canavandl) [inline bokeh/models/mappers.py]: @bryevdv - do you want me to add it back or deprecate? I have no strong opinion either way.

**Comment by bryevdv** [comment]
trailing whitespace: 
```
E       AssertionError: Code quality issues:
E         File contains trailing whitespace: bokeh/models/mappers.py, line 73.
E       assert 1 == 0
E        +  where 1 = len(['File contains trailing whitespace: bokeh/models/mappers.py, line 73.'])
```
I think we will all need to get in the habit of running at least some of the tests locally before pushing.

File: bokeh/models/mappers.py (hunk 2 above)
Comment (bryevdv) [inline bokeh/models/mappers.py]: let's just keep it for now

**Comment by jbednar** [comment]
Looks good to me now.

**Comment by bryevdv** [comment]
Yup, will merge when green.

**Comment by bryevdv** [comment]
> Maybe it would have been good to separate that into two PRs, then, one for the refactoring (no change to behavior) and one for the new functionality.

I agree that is probably a good ideal to strive for. I'm still learning the practice of patience :)

**Comment by bryevdv** [comment]
spurious failures, I think, restarted the two tests

**Comment by bryevdv** [comment]
OK it's the same test consistently failing... no idea what tho

File: bokeh/models/mappers.py [hunk 3]
```diff
@@ -5,10 +5,9 @@
 
 from ..model import Model
 from ..core.properties import abstract
-from ..core.properties import Float, Color, Enum, Seq, Instance
+from ..core.properties import Float, Color, Enum, Seq
 from ..core.enums import Palette
 from .. import palettes
-from .transforms import Jitter
```
Comment (bryevdv) [inline bokeh/models/mappers.py]: This is causing the test problem. The example does `from bokeh.models import Jitter` which was only "working: because this import was transitively available in `bokeh.models` @Maggie-M is going to make another issue/PR to fix, then you can merge that into this PR and it will pass.

**Comment by bryevdv** [comment]
left line comment about test failure. Will have a fix PR to merge into this soon.

**Comment by mattpap** [comment]
@canavandl, @bryevdv, according to MDN, `Math.log1p()` isn't available in IE (any version). A polyfill is needed.

**Comment by jbednar** [comment]
@mattpap, thanks!
